parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/FeatureAnalysers.scala

```diff
@@ -0,0 +1,224 @@
+      Seq(
+        Feature[String](digits)(digitsAnalyser()),
+        Feature[Root[_]](word)(wordAnalyser()),
+        Feature[String](length)(lengthAnalyser()),
+        Feature[Boolean](ordinal)(ordinalAnalyser()),
+        Feature[Boolean](endsInPunctuation)(endsInPunctuationAnalyser()),
+        Feature[Boolean](directional)(directionalAnalyser()),
+        Feature[Boolean](outcode)(outcodeAnalyser()),
+        Feature[Boolean](postTown)(postTownAnalyser()),
+        Feature[Boolean](hasVowels)(hasVowelsAnalyser()),
+        Feature[Boolean](flat)(flatAnalyser()),
+        Feature[Boolean](company)(companyAnalyser()),
+        Feature[Boolean](road)(roadAnalyser()),
+        Feature[Boolean](residential)(residentialAnalyser()),
+        Feature[Boolean](business)(businessAnalyser()),
+        Feature[Boolean](locational)(locationalAnalyser()),
+        Feature[Int](hyphenations)(hyphenationsAnalyser())
+      ).sortBy(_.name):_*
+    )(
+      //TODO Impl aggr feature to crfJniInput
+      FeatureAggregate[Boolean](rawStringStart)(rawStringStartAggrAnalyser()),
+      FeatureAggregate[Boolean](rawStringEnd)(rawStringEndAggrAnalyser()),
+      FeatureAggregate[Boolean](singleton)(singletonAggr())
+    )
+  }
+
+  object ADT {
+    trait Root[T] extends CrfType[T] {
+      def value: T
+    }
+    implicit class MyString(override val value: String) extends Root[String]
+    implicit class MyBoolean(override val value: Boolean) extends Root[Boolean]
+  }
+
+
+  //TODO scaladoc
+  val rawStringStart = "rawstring.start"
+  def rawStringStartAggrAnalyser(): CrfAggregateFeatureAnalyser[Boolean] = IndexAggregateFeatureAnalyser(_ => 0)
+
+  val rawStringEnd = "rawstring.end"
+  def rawStringEndAggrAnalyser(): CrfAggregateFeatureAnalyser[Boolean] = IndexAggregateFeatureAnalyser(_.length - 1)
+
+  val singleton = "singleton"
+  def singletonAggr(): CrfAggregateFeatureAnalyser[Boolean] = CrfAggregateFeatureAnalyser[Boolean] { (tokens, token) =>
+    tokens.length == 1 && tokens(0) == token
+  }
+
+  val word: FeatureName = "word"
+  /**
+    * @return true if the string is all digits, false if not
+    */
+  def wordAnalyser(): CrfFeatureAnalyser[Root[_]] = CrfFeatureAnalyser[Root[_]] { str =>
+    str.allDigits[Root[_]] { isAllDigits =>
+      if(isAllDigits) {
+        false
+      } else {
+        str
+      }
+    }
+  }
+
+  val length: FeatureName = "length"
+  /**
+    * @return the length of the string as a string
+    */
+  def lengthAnalyser(): CrfFeatureAnalyser[String] = CrfFeatureAnalyser[String] { str =>
+    val length = str.length
+    str.allDigits[String] { isAllDigits =>
+      if(isAllDigits) {
+        s"d:$length"
+      } else {
+        s"w:$length"
+      }
+    }
+  }
+
+  val endsInPunctuation: FeatureName = "endsinpunc"
+  /**
+    * @return true if the last character of the string is a '.', false if not
+    */
+  def endsInPunctuationAnalyser(): CrfFeatureAnalyser[Boolean] = CrfFeatureAnalyser[Boolean](_.last == '.')
```

Comment:
A lot of these accessors are not modifiers and have no side effect so you can remove `()`.